src/front/glsl/functions.rs

```diff
@@ -24,13 +46,41 @@ impl Program<'_> {
         match fc {
             FunctionCallKind::TypeConstructor(ty) => {
                 let h = if args.len() == 1 {
-                    let is_vec = match *self.resolve_type(ctx, args[0].0, args[0].1)? {
-                        TypeInner::Vector { .. } => true,
-                        _ => false,
+                    let expr_type = self.resolve_type(ctx, args[0].0, args[0].1)?;
+
+                    let vector_size = match *expr_type {
+                        TypeInner::Vector{ size, .. } => Some(size),
+                        _ => None,
                     };
 
+                    // Special case: if casting from a bool, we need to use Select and not As.
+                    let result_type_inner = &self.module.types[ty].inner;
+                    let result_scalar_kind = result_type_inner.scalar_kind();
+                    if expr_type.scalar_kind() == Some(ScalarKind::Bool) && result_scalar_kind.is_some() && result_scalar_kind != Some(ScalarKind::Bool) {
+                        let c0 = self.add_constant_value(result_scalar_kind.unwrap(), 0u64);
```

Comment:
A dirty way to simplify this would be `let result_scalar_kind = result_type_inner.scalar_kind().unwrap_or(ScalarKind::Bool);`

A pure/clean way would be to express this as a `match`:
```rust
match result_type_inner.scalar_kind() {
  Some(result_scalar_kind) if expr_type.scalar_kind() == Some(ScalarKind::Bool) => {...}
  _ => {}
}
```

Reply:
Done.